cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp

```diff
@@ -102,15 +130,49 @@ inline sycl::event prepare_correlation(sycl::queue& q,
     });
 }
 
+template <typename Float>
+inline sycl::event finalize_correlation_with_covariance(sycl::queue& q,
+                                                        std::int64_t row_count,
+                                                        const ndview<Float, 2>& cov,
+                                                        const ndview<Float, 1>& tmp,
+                                                        ndview<Float, 2>& corr,
+                                                        const event_vector& deps) {
+    const auto n = row_count;
+    const auto p = cov.get_dimension(1);
+    const Float inv_n1 = (n > 1.0f) ? Float(1.0 / double(n - 1)) : 1.0f;
+    const Float* tmp_ptr = tmp.get_mutable_data();
+    Float* corr_ptr = corr.get_mutable_data();
+    Float* cov_ptr = cov.get_mutable_data();
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = make_range_2d(p, p);
+
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::item<2> id) {
+            const std::int64_t gi = id.get_linear_id();
+            const std::int64_t i = id.get_id(0);
+            const std::int64_t j = id.get_id(1);
+
+            if (i < p && j < p) {
```

Comment:
Why do you need this check? You have already specified `range(p, p)` so I see no reason to check it yet another one time